File: topics/Graph/CF292B.py
N,M = map(int,input().split())
dig = [0 for _ in range(N)]
for x,y in [map(int,input().split() for _ in range(m)]:
        dig[x-1],dig[y-1] = dig[x-1]+1,dig[y-1]+1
digs = set(dig)
r = "unknown"
#这里比较巧妙,不是将每个点的degree进行统计,而是直接统计有什么可能的,{1,2}的情形的确只有一种
if digs =={2}:r = "ring"
elif digs=={1,2}:r ="bus"
else digs=={1,N-1}r = "star"
print(r,"topology")

File: topics/Graph/CF_training/CF292B.cpp
/*  -*- coding: utf-8 -*- */
#include <bits/stdc++.h>
#define MEM(a,num) memset(a,num,sizeof(a))
#define rep(i,a,b) for(int i=a;i<=b;i++)
#define fi first
#define pb push_back
#define se second
using namespace std;
typedef long long ll;
typedef pair<int,int> PP;
//const double esp=1e-6
const int maxn  = 1e5+100;
int sz[maxn],N,M;
int main()
{
	scanf("%d%d",&N,&M);
	MEM(sz,0);
	vector<int> cnt;cnt.resize(N);
	int a,b;
	for(int i=1;i<=M;i++){
		scanf("%d%d",&a,&b);a--,b--;
		cnt[a]++,cnt[b]++;
	}
	sort(cnt.begin(),cnt.end());
	for(int i=0;i<N;i++)sz[cnt[i]]++;
	if(sz[2]==N)printf("ring topology\n");
	else if(sz[1]==2&&sz[2]==N-2)printf("bus topology\n");
	else if(sz[1]==N-1&&sz[N-1]==1)printf("star topology\n");
	else printf("unknown topology\n");
}

